graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -32,4 +40,187 @@ def lower_ir(ir_blocks, query_metadata_table, type_equivalence_hints=None):
     Returns:
         tree representation of IR blocks for recursive traversal by SQL backend.
     """
-    raise NotImplementedError(u'SQL IR lowering is not yet implemented.')
+    _validate_all_blocks_supported(ir_blocks, query_metadata_table)
+    construct_result = ir_blocks.pop()
+    query_path_to_location_info = _map_query_path_to_location_info(query_metadata_table)
+    query_path_to_output_fields = _map_query_path_to_outputs(
+        construct_result, query_path_to_location_info)
+
+    # iteratively construct SqlTree
+    query_path_to_node = {}
+    block_index_to_location = _map_block_index_to_location(ir_blocks)
+    tree_root = None
+    for index, block in enumerate(ir_blocks):
+        if isinstance(block, constants.SKIPPABLE_BLOCK_TYPES):
+            continue
+        location = block_index_to_location[index]
+        if isinstance(block, (blocks.QueryRoot,)):
+            query_path = location.query_path
+            if tree_root is not None:
+                raise AssertionError(
+                    u'Encountered QueryRoot {} but tree root is already set to {} during '
+                    u'construction of SQL query tree for IR blocks {} with query '
+                    u'metadata table {}'.format(
+                        block, tree_root, ir_blocks, query_metadata_table))
+            tree_root = SqlNode(block=block, query_path=query_path)
+            query_path_to_node[query_path] = tree_root
+        else:
+            raise AssertionError(
+                u'Unsupported block {} unexpectedly passed validation for IR blocks '
+                u'{} with query metadata table {} .'.format(block, ir_blocks, query_metadata_table))
+
+    return SqlQueryTree(tree_root, query_path_to_location_info, query_path_to_output_fields)
+
+
+def _validate_all_blocks_supported(ir_blocks, query_metadata_table):
+    """Validate that all IR blocks and ConstructResult fields passed to the are supported.
+
+    Args:
+        ir_blocks: List[BasicBlock], IR blocks to validate.
+        query_metadata_table: QueryMetadataTable, object containing all metadata collected during
+                              query processing, including location metadata (e.g. which locations
+                              are folded or optional).
+
+    Returns:
+        None
+
+    Raises:
+        NotImplementedError, if any block or ConstructResult field is unsupported.
+    """
+    if len(ir_blocks) < 3:
+        raise AssertionError(
+            u'Unexpectedly attempting to validate IR blocks with fewer than 3 blocks. A minimal '
+            u'query is expected to have at least a QueryRoot, GlobalOperationsStart, and '
+            u'ConstructResult block. The query metadata table is {}.'.format(query_metadata_table))
+    construct_result = ir_blocks[-1]
```

Comment:
I wouldn't name this variable `construct_result` until you've checked its type. Otherwise, a future edit to this file may add some functionality between lines 95 and 105 that assumes its type is `ConstructResult`, which will break before the assertion catches it and therefore be difficult to debug.